### ISSUE TITLE
Raspbee

### ISSUE THREAD

**Issue by snylee** [riverloopsec/killerbee#201]
Hi Guys

I'm kinda new to this so please bare with me.
I'm a student in my final year of university. For my final year project I have chosen to focus on the penetration of Zigbee networks/devices. 
So far I have purchased a Raspberry Pi with the Raspbee Premium module. I am able to see traffic through Wireshark using software from Bishop Fox's software, ZigDiggity. But i want to do more, which is how I've come across Killerbee.
I've noticed that the hardware requirements for Killerbee do not include that of Raspbee. Do you know if there is anyway I could use Killerbee with my Raspbee module? I only ask this because most of the hardware required for Killerbee are no longer in stock, especially the Atmel RZUSB Stick. I would need to sniff and inject.
Any ideas would be mostly beneficial.

Thanks Guys
Simon

**Comment by rmspeers** [comment]
Yes, you should be able to add Raspbee support relatively simply -- copy the pattern in another dev_*.py file and implement those functions that are in there to speak to the Raspbee hardware via it's serial interface. Your implementation likely could be informed by looking at `zigdiggity/radios/raspbee_radio.py` or you can implement using the Raspbee specifications.

**Comment by snylee** [comment]
Thanks for the quick reply rmspeers. 
So please let me know if i've got this right. 
-create a new .py script called dev_raspbee.py -(for example)
-copy the script from "zigdiggity/radios/raspbee_radio.py" into my new dev_raspbee.py script?
I'm a bit confused. Ha

**Comment by rmspeers** [comment]
You will want to create dev_raspbee.py, and can base it on dev_apimote.py as an example. However, you can’t use copy something in - you will need to implement those functions (pnext, set_channel, etc) based on the Raspbee communication protocol (which, as I mentioned,  "zigdiggity/radios/raspbee_radio.py” may be one reference for how to do if you wanted to read it).

You’ll need to implement those methods and debug from that point to ensure you’re getting all the data in the expected format (e.g., FCS prepended vs null’ed, etc).

**Comment by snylee** [comment]
Damn. So I’m not a programmer so i have no clue how to implement it across or what to look for in the zigdiggity/radios/raspbee_radio.py script either. Any other ideas to add support? Or do you have any idea of where i could get one of the supported hardware radios from?

**Comment by rmspeers** [comment]
Code needs to be written for there to be support for a radio. I've started a branch, tagged 201-raspbee-support, but have not had time to write most of it or test it - it serves only as an example to be built upon and tested.

**Comment by rmspeers** [comment]
For supported radios, you should be able to get the SiLabs Node Test supported boards from DigiKey and similar distributors easily, if those support the features you need. The ApiMote v4beta you can get from Attify who resells them. TelosB board clones I believe are readily available in the EU as well, and can be shipped internationally.

**Comment by snylee** [comment]
Ok thank for. I appreciate your time and effort. In the meantime time trying to do a quick corse on phthon. I know the basics but im not there yet. 😁

**Comment by wang70880** [comment]
What's the status of Raspbee firmware development?

**Comment by rmspeers** [comment]
At this time, I have not had time to dedicate to writing a Raspbee firmware. We welcome any contributions of this firmware.

**Comment by wang70880** [comment]
I would like to help develop the Raspbee firmware. Do you have any idea to help quickly get started? Actually I am confused. 

In my view, Raspbee needs its firmware which provides some basic APIs through UART. KillerBee on the other side is more likely a toolkit which is developed by using these APIs provided by Raspbee firmware.

Should I also design Raspbee firmware, or use its orginal firmware? In my view, the dev_raspbee.py is just a software abstraction of Raspbee, and it didn't change the firmware of it.

**Comment by rmspeers** [comment]
See above notes on https://github.com/riverloopsec/killerbee/issues/201#issuecomment-583901216 on what you need to do for the KillerBee "driver" layer. You can use the default firmware if you wish, see above considerations. `dev_raspbee.py` is, as you say, the abstraction layer for speaking to the firmware. I would start with implementing that to talk to the default firmware and you can update firmware if you find things are missing after doing that which you need.

**Comment by wang70880** [comment]
> See above notes on [#201 (comment)](https://github.com/riverloopsec/killerbee/issues/201#issuecomment-583901216) on what you need to do for the KillerBee "driver" layer. You can use the default firmware if you wish, see above considerations. `dev_raspbee.py` is, as you say, the abstraction layer for speaking to the firmware. I would start with implementing that to talk to the default firmware and you can update firmware if you find things are missing after doing that which you need.

Thanks, and it seems that my understanding is correct. The last question is that I don't know how Zigdiggty found that CMD_SET_CHANNEL = 0x35 or CMD_RX = 0x31. To be honest, I didn't find any documentation about the communication protocol with Raspbee firmware. Do you have any idea?

**Comment by wang70880** [comment]
However, I found that the firmware of Raspbee is not open-sourced, and the functionality of original firmware is limited, e.g., changing PANID. What should we do to solve this firmware problem?
